### ISSUE TITLE
Got iotedge-agent AMD64 image when trying to pull linux-arm32-v7

### ISSUE THREAD

**Issue by FluxReality** [Azure/iotedge#951]
Hi,

Can someone please explain the output for this command:
` docker manifest inspect -v mcr.microsoft.com/azureiotedge-agent:1.0.4-linux-arm32v7 | jq .Descriptor.platform`

On my arm linux device. Gives the folowing result:

`
{
  "architecture": "amd64",
  "os": "linux"
}
`

Shouldn't this be architecture **arm32v7** or something?

The reason I'm asking is that I now get the following error in my syslog when iotedged is trying to start:

> Mar 13 13:19:45 tegra-ubuntu dockerd[2068]: time="2019-03-13T13:19:45.614820517+01:00" level=info msg="Attempting next endpoint for pull after error: no matching manifest for unknown in the manifest list entries"
>
> Mar 13 13:19:45 tegra-ubuntu iotedged[2070]: 2019-03-13T12:19:45Z [INFO] - [mgmt] - - - [2019-03-13 12:19:45.627189984 UTC] "PUT /modules/edgeAgent?api-version=2018-06-28&start=True HTTP/1.1" 404 Not Found 33 "-" "-" pid(1292)

Not letting the iotedge daemon start properly, even though this was working fine before.

**Comment by myagley** [comment]
We cross compile all of the arm platforms on an amd64 machine which is why the manifest for the arm images may be reporting amd64. However, this shouldn't affect anything as far as I understand it. The more interesting information is what the manifest shows for `mcr.microsoft.com/azureiotedge-agent:1.0.4`

This should show three entries for `linux/amd64`, `linux/arm32`, and `windows/amd64`.

What tag are you using in the deployment for the Edge Agent? If you want version `1.0.4`, you should be using the `1.0.4` tag with no platform information.

**Comment by FluxReality** [comment]
I am using this tag, because I'm trying to deploy on a Nvidia Jetson TX2 device which is arm64v8.

Using the **mcr.microsoft.com/azureiotedge-agent:1.0.4-linux-arm32v7** did work, until it didn't... (i restarted the iotedged after removing both images from the local store).

**Comment by FluxReality** [comment]
Also @myagley if you do:

`docker manifest inspect -v mcr.microsoft.com/azureiotedge-agent:1.0.5-linux-arm32v7 | jq .Descriptor.platform`

You get the expected result:

`
{
  "architecture": "arm",
  "os": "linux"
}
`

I'm not sure that the **architecture** key describes the architecture it was built _ON_ but rather what it was built _FOR_ (someone may be able to point me to the documentation on this).

Hence giving problems when trying to pull on an ARM architecture and getting an AMD image.

**Comment by FluxReality** [comment]
For refference I have previously been folowing the issue #104 and [theese](https://devblogs.microsoft.com/iotdev/a-workaround-to-run-azure-iot-edge-on-arm64-devices/) instructions on how to get IoT edge working on Nvidia Jetson device (and it was working until recently when i restarted the daemon).

**Comment by arsing** [comment]
The platform should be the arch it was built for, not what it was built on. It comes from the manifest template that we provide.

Also:

```sh
$ docker manifest inspect -v mcr.microsoft.com/azureiotedge-agent:1.0.4 | jq 'map({ digest: .Descriptor.digest, platform: .Descriptor.platform })'

[
  {
    "digest": "sha256:5c1f772c1ca0f384daadc2bffeff20c2490614ebfc65ef9fc3bf69bde052b526",
    "platform": {
      "architecture": "amd64",
      "os": "linux"
    }
  },
  {
    "digest": "sha256:86854025ed40f8fe4b20a59e189ee549efe4fcfecec23cc3a7c3125c52233e21",
    "platform": {
      "architecture": "arm",
      "os": "linux"
    }
  },
  {
    "digest": "sha256:3293e4290fc85a7cdc358545bee528a38557ac1992280a871518262b39ffcab4",
    "platform": {
      "architecture": "amd64",
      "os": "windows",
      "os.version": "10.0.17134.228"
    }
  }
]

$ docker manifest inspect -v mcr.microsoft.com/azureiotedge-agent:1.0.4-linux-arm32v7 | jq '{ digest: .Descriptor.digest, platform: .Descriptor.platform }'

{
  "digest": "sha256:86854025ed40f8fe4b20a59e189ee549efe4fcfecec23cc3a7c3125c52233e21",
  "platform": {
    "architecture": "amd64",
    "os": "linux"
  }
}
```

So the platform is `arm` when querying all the images, but `amd64` when querying that specific image. I'm inclined to believe it's an ACR bug.

**Comment by arsing** [comment]
@FluxReality We can follow up with ACR to see if there's some problem with 1.0.4's metadata on their end. But can you update to 1.0.6 to resolve your situation?

**Comment by FluxReality** [comment]
@arsing Yes I did, and it's been resolved. Thankyou.

**Comment by arsing** [comment]
Thanks. I'll close this once we figure out what the issue was from MCR. (It also affects all the older images 1.0.0 through 1.0.3, so perhaps something got fixed between the 1.0.4 and 1.0.5 releases and wasn't applied retroactively.)

**Comment by arsing** [comment]
So the arch in the multi-arch manifest is independent of the arch for the images themselves. The output I gave above is to be expected if the manifest declares the image to have `architecture: arm` but the image is annotated with `architecture: amd64`. Indeed, by querying the ACR info separately, I can see that the image is tagged with `architecture: amd64`.

The image's architecture annotation is supposed to be set automatically by Docker when building and pushing the image, based on the base image of the image. The ACR folks mentioned that Docker used to have a bug where it would annotate the image architecture based on the machine used to build the image instead. It's possible the VSTS VMs (hosted Linux pool) we used to build 1.0.4 and earlier had this old version of Docker, and the ones we used for 1.0.5+ have the new version with the bugfix.

Unfortunately there is no way to edit the images in ACR now to set the correct architecture. We would have to build and deploy new images, plus a new manifest pointing to those new images. It doesn't seem to be worth it to me since new images are fine, so I'll close this.
